### ISSUE TITLE
False error about destructuring assignments when using _ as implicit array length

### ISSUE THREAD

**Issue by tesuji** [rust-lang/rust#81268]
I tried this code:
([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8d4d7c74fe0c8eb9979f24254c603167))

```rust
pub fn foo() {
    let s: [u8; 10];
    s = [0; _];
    dbg!(s);
}

```

I expect to see this happen: There are no errors complain about destructuring assignments are unstable.
Instead: The compiler complains about that:

Errors:

```rust
   Compiling playground v0.0.1 (/playground)
error[E0658]: destructuring assignments are unstable
 --> src/lib.rs:3:13
  |
3 |     s = [0; _];
  |             ^
  |
  = note: see issue #71126 <https://github.com/rust-lang/rust/issues/71126> for more information
  = help: add `#![feature(destructuring_assignment)]` to the crate attributes to enable

error: in expressions, `_` can only be used on the left-hand side of an assignment
 --> src/lib.rs:3:13
  |
3 |     s = [0; _];
  |             ^ `_` not allowed here
```

Build using the Nightly version: 1.51.0-nightly (2021-01-21 202720bf483088dbdb34)

@rustbot label C-bug F-destructuring_assignment D-incorrect

**Comment by varkor** [comment]
I'm not sure this is necessarily incorrect (it's part of the `destructuring_assignment` feature gate, and it is only allowed on the left-hand side of an assignment). I'm not sure what better diagnostic we could be emitting here: any suggestions?

**Comment by ChayimFriedman2** [comment]
Even after destructuring assignments were stabilized, [it still says that](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f0dac69c1901f59fe09298f8425fdc2e):
```
error: in expressions, `_` can only be used on the left-hand side of an assignment
 --> src/main.rs:2:27
  |
2 |     let _: [u16; 1] = [0; _];
  |                           ^ `_` not allowed here
```
In addition to the correct error
```
error[[E0658]](https://doc.rust-lang.org/nightly/error-index.html#E0658): using `_` for array lengths is unstable
 --> src/main.rs:2:27
  |
2 |     let _: [u16; 1] = [0; _];
  |                           ^
  |
  = note: [see issue #85077 <https://github.com/rust-lang/rust/issues/85077>](https://github.com/rust-lang/rust/issues/85077) for more information
  = help: [add `#![feature(generic_arg_infer)]`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#) to the crate attributes to enable
```
Somewhat strangely, [it does work if I enable the feature](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=0bbce27f10c5bfa96911f82b2bd62fc8).

**Comment by estebank** [comment]
Current output:
```
error: in expressions, `_` can only be used on the left-hand side of an assignment
 --> src/lib.rs:3:13
  |
3 |     s = [0; _];
  |             ^ `_` not allowed here

error[E0658]: using `_` for array lengths is unstable
 --> src/lib.rs:3:13
  |
3 |     s = [0; _];
  |             ^
  |
  = note: see issue #85077 <https://github.com/rust-lang/rust/issues/85077> for more information
  = help: add `#![feature(generic_arg_infer)]` to the crate attributes to enable
```
